Expose group/ungroup buttons in site editor

ntsekouras opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
Fixes: https://github.com/WordPress/gutenberg/issues/25687

This PR just adds the group/ungroup buttons in Site Editor.

ntsekouras committed:
use group/ungroup buttons

ntsekouras committed:
unrelated typo fix

gziolo commented:
I find it surprising that you had to export the component that adds grouping actions to the menu. How it works for edit post page without exporting? What about edit widgets and edit navigation screens? It feels like it should be a built-in feature in the shared editor, similar to how it adds actions for reusable blocks. @youknowriad, does it make sense?

ntsekouras commented:
>How it works for edit post page without exporting?

It's used only there and is an internal dependency..

ellatrix commented:
I agree, should be built in imo.

gziolo reviewed (changes requested):
It creates cyclic dependencies between two modified packages. There should be a slot in the editor that allows injecting menu items in the block editor. Reusable blocks should be configured this way.

youknowriad commented:
I agree we could move this to `block-editor` directly if a "grouping block" is available. What's stopping us from doing so?

ntsekouras committed:
move ConvertToGroupButtons to block-editor

gziolo commented on packages/block-editor/src/components/block-settings-menu/block-settings-dropdown.js:
```diff
@@ -168,6 +169,7 @@ export function BlockSettingsDropdown( {
 									/>
 								) }
 							</MenuGroup>
+							<ConvertToGroupButton onClose={ onClose } />
```
Noting that `ConvertToGroupButton` internally uses `BlockSettingsMenuControls` fill that is going to render this menu item inside `BlockSettingsMenuControls.Slot` slot. So no matter where you put it in the rendering tree, it will get placed somewhere else. It's very confusing, to be honest :)

youknowriad replied:
Now that it's in the block-editor, we can remove the slot usage

gziolo replied:
There are also reusable blocks that use it. I think this is why this slot was introduced in the first place. When searching for `<BlockSettingsMenuControls />` use I can also see other functionalities injected with it, with `PluginBlockSettingsMenuItem` exposed from `wp.editPost`.

ntsekouras replied:
Yeah I don't think the slot should be removed. I'll check it today better.

gziolo commented on packages/editor/src/components/provider/index.js:
```diff
@@ -120,7 +119,6 @@ function EditorProvider( {
 					>
 						{ children }
 						<ReusableBlocksMenuItems />
-						<ConvertToGroupButtons />
```
Hmm, so it was placed next to reusable blocks in the menu already. Now, I'm curious why it isn't shared between edit post, edit site, and edit widgets?

ntsekouras replied:
Actually reusable blocks are used in editor, widgets and there is a similar but custom implementation in site-editor that 'maps' reusable blocks -> template parts. So for reusables might make sense not be in all editors.

I'll investigate today for a solution of this, but I'm curious why `convert to group` wasn't in `block-editor` from the beginning.

youknowriad replied:
it wasn't in block-editor because defining a default  group block was added after the group block was added.

ntsekouras replied:
Thanks for explaining 👍

ntsekouras committed:
add ConvertToGroupButton to blockSettingsDropdown

ntsekouras committed:
remove previous component from 'editor'

gziolo commented on packages/block-editor/src/components/block-settings-menu-controls/index.js:
```diff
@@ -31,9 +36,14 @@ const BlockSettingsMenuControlsSlot = ( { fillProps, clientIds = null } ) => {
+			{ ( fills ) => {
+				return (
+					<MenuGroup>
+						{ ! isEmpty( fills ) && fills }
```
It's engough to have now:
```suggestion
						{ fills }
```

This check existed to avoid rendering the menu group with no content inside.

gziolo commented on packages/block-editor/src/components/block-settings-menu-controls/index.js:
```diff
@@ -31,9 +36,14 @@ const BlockSettingsMenuControlsSlot = ( { fillProps, clientIds = null } ) => {
+				return (
+					<MenuGroup>
+						{ ! isEmpty( fills ) && fills }
+						<ConvertToGroupButton />
```
Do we want to have this option always displayed as the last one? There might be several custom items injected by plugin authors. We should consider having it before `fills`, but it would also mean that it would flip the order with reusable blocks.

ntsekouras replied:
That's why I placed it below, because of the `reusable` order.. I don't have a strong opinion on this and don't have any context if the order was discussed in the past

gziolo commented:
This PR is very close now that the latest changes were applied. The failing e2e tests seem to be related to the updated order of items in the block settings menu. I guess it's fine to change the order but we will need to update tests to align with that.

ntsekouras committed:
pass onClose to ConvertToGroupButton

ntsekouras commented:
@gziolo the order hasn't changed. I missed passing the `onClose` so it was a valid failure because of that. Addressed the feedback as well. Noting though that the previous component using `withSelect` and `withDispatch` and also refactored to use hooks.

youknowriad commented on packages/block-editor/src/components/block-settings-menu-controls/index.js:
```diff
@@ -31,9 +36,14 @@ const BlockSettingsMenuControlsSlot = ( { fillProps, clientIds = null } ) => {
+				return (
+					<MenuGroup>
+						{ fills }
+						<ConvertToGroupButton onClose={ fillProps?.onClose } />
```
Did you consider adding it directly to `BlockSettingsDropdown` ? I guess it might change its position a little bit but maybe that's fine?

ntsekouras replied:
Yes, but previously was in the same `MenuGroup` with the other fills and that's the reason I placed it here.

Alternatively I should create a new `MenuGroup` for `ConvertToGroupButton ` and add css to remove the border. I have the feeling that we generally don't want t0o many `borders` in such places (another example is Toolbar but don't know if it applies the same) 🤔

youknowriad replied:
Sounds good to me

gziolo reviewed (approved):
Code wise 🚢

gziolo commented:
@jasmussen, do you think that order matters that much in the block settings menu?

<img width="236" alt="Screen Shot 2021-01-14 at 15 02 55" src="https://user-images.githubusercontent.com/699132/104600739-b0ab9b00-5679-11eb-97bd-6698e9d1a398.png">

With this PR, `Group` & `Ungroup` will be always placed as last in the group. It's going be after any item registered by plugins.

ntsekouras commented:
>With this PR, Group & Ungroup will be always placed as last in the group. It's going to be placed after any item registered by plugins.

Also the alternative will be for it to be always on top. Above `reusable, make template part(site-editor)` etc...

jasmussen commented:
> With this PR, Group & Ungroup will be always placed as last in the group. It's going to be placed after any item registered by plugins.

Seems okay to me.
